### PR TITLE
New VRMLLoader based on the VrmlParser project

### PR DESCRIPTION
**Pull request by bartmcleod**
By using the VrmlParser project dependencies as a single minified script, the VRMLLoader can now be used without external dependencies.

This allows users to easily and reliably load VRML models with ThreeJs while adding better support for VRML can be managed separately.

### CODE REVIEW

**Comment by mrdoob** [comment]
Thanks!

File: examples/webgl_loader_vrml.html [hunk 1]
```diff
@@ -1,129 +1,214 @@
...
-			animate();
-
-			function init() {
-
-				camera = new THREE.PerspectiveCamera( 60, window.innerWidth / window.innerHeight, 0.01, 1e10 );
-				camera.position.z = 6;
+		}; // key value store of cameras based on VRML viewpoint nodes, stored by their name.
 
-				controls = new THREE.OrbitControls( camera );
+		var vrmlLoader = new THREE.VRMLLoader(vrmlParser, new THREE.LoadingManager(), true);
```
Comment (mrdoob) [inline examples/webgl_loader_vrml.html]: Where is `vrmlParser` defined?

**Comment by mrdoob** [comment]
Hmm, I may need to revert this. This new loader broke VRML support in https://threejs.org/editor/ (dev)... 

As you can see, all the loaders follow the same kind of interface:
https://github.com/mrdoob/three.js/blob/dev/editor/js/Loader.js#L28-L415

**Comment by bartmcleod** [comment]
I wasn't aware of the interface. Like I said before, many different integrations can be thought of. `vrmlParser` is defined in the output of the PEG.js grammar. It can be changed it necessary. It is okay if you revert, than I will look into an integration that conforms to the interface. I do not agree in general with putting everything in a single file. Maybe that is a best practice in the JavaScript world? It certainly isn't in other languages. Parsing VRML is very complex, especially if you look at the compliance rules in the specification. Managing that within a single file will be a total nightmare.

However, it should be possible to hide all the nitty gritty behind an interface, so I will look at the interface and see what I can do.

**Comment by mrdoob** [comment]
> I do not agree in general with putting everything in a single file.

That's not what I meant. What I meant was that currently you're passing `VrmlParser` to the `VRMLLoader`. So the user needs to be aware of both `VrmlParser` and `VRMLLoader`. What I'm suggesting is that it would be better if `VRMLLoader` interacted with `VrmlParser` internally.

Something like this:

```js
	parse: function (data) {

		if ( 'VrmlParser' in window === undefined ) {

			console.error( 'THREE.VRMLLoader: Couldn't find VrmlParser' );
			return; 

		}

		var parser = new VrmlParser();

		try {
			var tree = parser.parse(data);

			this.log(tree);
			return tree;
		} catch ( e ) {
			this.log('Exception with message ' + e.message);

			if ( undefined !== e.location ) {
				this.log('Exception at location start: offset: ' + e.location.start.offset + ' line: ' + e.location.start.line + ' column: ' + e.location.start.column);
				this.log('Exception at location end: offset: ' + e.location.end.offset + ' line: ' + e.location.end.line + ' column: ' + e.location.end.column);
			}

			return;
		}
	},
```

**Comment by bartmcleod** [comment]
@mrdoob  After lots and lots of refactoring, I just wanted to drop a quick note here,  to let you know I did not abandon the idea of a loading mechanism that adheres to the existing interface. It is pretty hard however, to get it working, because the interaction with the viewpoints kind of pulls everything into the loader. This makes it inflexible. Is it ok to load some dependencies in the constructor of the loader? I do not think any particular interface applies to the loader constructor does it?

**Comment by bartmcleod** [comment]
@mrdoob New attempt can be found here: https://github.com/mrdoob/three.js/pull/10371

**Pull request merged**